pisa/stages/xsec/dis_sys.py

```diff
@@ -0,0 +1,213 @@
+            nu = 'Nu' if container['nubar'] > 0 else 'NuBar'
+
+            lgE = np.log10(container['true_energy'].get('host'))
+            bjorken_y = container['bjorken_y'].get('host')
+            dis = container['dis'].get('host')
+
+            #
+            # Calculate variation of total cross section
+            #
+
+            lgE_min = 2. if self.extrapolation_type == 'constant' else 1.68
+            mask = lgE <= lgE_min
+
+            w_tot = np.ones_like(lgE)
+            
+            poly_coef = extrap_dict[nu][current]['poly_coef']
+            w_tot[~mask] = np.polyval(poly_coef, lgE[~mask])
+           
+            if self.extrapolation_type == 'constant':
+                poly_coef = extrap_dict[nu][current]['poly_coef']
+                w_tot[mask] = np.polyval(poly_coef, lgE_min * w_tot[mask]) 
+            elif self.extrapolation_type == 'linear':
+                lin_coef = extrap_dict[nu][current]['linear']
+                w_tot[mask] = np.polyval(lin_coef, lgE[mask])
+            elif self.extrapolation_type == 'higher':
+                poly_coef = extrap_dict[nu][current]['poly_coef']
+                w_tot[mask] = np.polyval(poly_coef, lgE[mask]) 
+            else:
+                raise ValueError('Unknown extrapolation type "%s"'%self.extrapolation_type)
+
+            # mke centered arround 0, and set to 0 for all non-DIS events
+            w_tot -= 1
+            w_tot *= dis
+          
+            container["dis_correction_total"] = w_tot
+
+            #
+            # Calculate variation of differential cross section
+            #
+
+            lgE_min = 2.
+            mask = lgE <= lgE_min
+
+            w_diff = np.ones_like(lgE)
+
+            if current == 'CC' and container['nubar'] > 0:
+                weight_func = wf_nucc
+            elif current == 'CC' and container['nubar'] < 0:
+                weight_func = wf_nubarcc
+            elif current == 'NC' and container['nubar'] > 0:
+                weight_func = wf_nunc
+            elif current == 'NC' and container['nubar'] < 0:
+                weight_func = wf_nubarnc
+
+            w_diff[~mask] = weight_func.ev(lgE[~mask], bjorken_y[~mask])
+            w_diff[mask] = weight_func.ev(w_diff[mask] * lgE_min, bjorken_y[mask])
+
+            # mke centered arround 0, and set to 0 for all non-DIS events
+            w_diff -= 1
+            w_diff *= dis
+            
+            container["dis_correction_diff"] = w_diff
+         
+
+    @profile
+    def apply_function(self):
+        dis_csms = self.params.dis_csms.m_as('dimensionless')
+
+        for container in self.data:
+            apply_dis_sys(
+                container['dis_correction_total'].get(WHERE),
+                container['dis_correction_diff'].get(WHERE),
+                dis_csms,
+                out=container['weights'].get(WHERE),
+            )
+            container['weights'].mark_changed(WHERE)
+
+
+if FTYPE == np.float64:
```

Comment:
Avoiding copy-paste errors got easier since f-strings; I'd like to propose that people use the following convention for these signatures from now on:
```python
FX = 'f8' if FTYPE == np.float64 else 'f4'

@guvectorize([f'({FX}, {FX}, {FX}, {FX}[:])'], '(),(),()->()', target=TARGET)
 def apply_dis_sys(...)
```